nexus/src/app/background/sync_service_zone_nat.rs

```diff
@@ -0,0 +1,315 @@
+                                    );
+                                    continue;
+                                },
+                            };
+
+                            let external_address =
+                                ipnetwork::Ipv4Network::new(external_ip, 32)
+                                    .unwrap();
+
+                            let nat_value = Ipv4NatValues {
+                                external_address: nexus_db_model::Ipv4Net(
+                                    omicron_common::api::external::Ipv4Net(
+                                        external_address,
+                                    ),
+                                ),
+                                first_port: MIN_PORT.into(),
+                                last_port: MAX_PORT.into(),
+                                sled_address: sled_address.into(),
+                                vni: nexus_db_model::Vni(nic.vni),
+                                mac: nexus_db_model::MacAddr(nic.mac),
+                            };
+
+                            // Append ipv4 nat entry
+                            ipv4_nat_values.push(nat_value);
+                        },
+                        // we explictly list all cases instead of using a wildcard,
+                        // that way if someone adds a new type to OmicronZoneType that
+                        // requires NAT, they must come here to update this logic as
+                        // well
+                        OmicronZoneType::Clickhouse {..} => continue,
+                        OmicronZoneType::ClickhouseKeeper {..} => continue,
+                        OmicronZoneType::CockroachDb {..} => continue,
+                        OmicronZoneType::Crucible {..} => continue,
+                        OmicronZoneType::CruciblePantry {..} => continue,
+                        OmicronZoneType::InternalNtp {..} => continue,
+                        OmicronZoneType::InternalDns {..} => continue,
+                        OmicronZoneType::Oximeter { ..} => continue,
+                    }
+                }
+            }
+
+            // if we make it this far this should not be empty:
+            // * nexus is running so we should at least have generated a nat value for it
+            // * nexus requies other services zones that require nat to come up first
+            if ipv4_nat_values.is_empty() {
+                error!(
+                    &log,
+                    "nexus is running but no service zone nat values could be generated from inventory";
+                );
+                return json!({
+                    "error": "nexus is running but no service zone nat values could be generated from inventory"
+                });
+            }
+
+            // reconcile service zone nat entries
+            let result = match self.datastore.ipv4_nat_sync_service_zones(opctx, &ipv4_nat_values).await {
```

Comment:
We discussed this in chat, but this whole flow has some implications for our calls to `ensure_nat_entry` / `ensure_ipv4_nat_entry`:

- If you're writing code in Nexus to make an **instance** with a NAT entry, you must call these functions explicitly. No one else will add your NAT entry to the DB!
- If you're writing code in Nexus to make a **service** with a NAT entry, you **should not call these functions**, and should rely on the "Nexus -> Sled Agent -> Inventory -> Service Zone NAT RPW" pathway" to ensure that these entries get populated. If you *tried* to add the entry to the DB, you'd risk a race condition between "collection via inventory / RPW" and "explicitly inserting the record".

I think that's okay, but we should add documentation around the `ensure_nat_entry` function to make that distinction extremely clear! I think it's kinda subtle that the same table is treated quite differently depending on the purpose of the NAT entry.

Reply:
Also, for posterity, I created this drawing to map out the data flow here.

My "TLDR" of the above is that I wanted to ensure we avoided having loops in this data flow graph:

![image](https://github.com/oxidecomputer/omicron/assets/3258857/370da577-2eb8-4f0a-81e9-91a0b0f6a98d)

Reply:
I think this is great feedback, adding some documentation to it now. I think one day we may get to a place where we can lift the NAT logic out of the service zone creation functionality in sled-agent, which could help us move towards a more consistent pattern of interacting with this table.